### PR TITLE
New tags: Copyright, ISRC, Artists, AlbumArtists, URL

### PR DESCRIPTION
**Pull request by JuniorJPDJ**
Also adds additional barcode tags to read from

Sources for tag mapping:
https://picard-docs.musicbrainz.org/en/appendices/tag_mapping.html#artists
https://wiki.hydrogenaud.io/index.php?title=Tag_Mapping
https://kodi.wiki/view/Music_tagging#albumartists

### CODE REVIEW

File: mediafile.py
Comment (JuniorJPDJ) [inline mediafile.py:1816]: Effect I wanted to achieve was reading data from this 3 tags and saving only BARCODE tag.
Is it a way to do this?

File: mediafile.py
Comment (JuniorJPDJ) [inline mediafile.py:1647]: How can I do something like this?
I've tag like:
```
'TXXX:ARTISTS': TXXX(encoding=<Encoding.UTF16: 1>, desc='ARTISTS', text=['Miuosh/Jimek/Narodowa Orkiestra Symfoniczna Polskiego Radia w Katowicach']),
```
And it's supposed to be interpreted as list: ['Miuosh', 'Jimek']

Comment (JuniorJPDJ) [reply mediafile.py:1647]: Source for this tag: https://picard-docs.musicbrainz.org/en/appendices/tag_mapping.html#artists

File: mediafile.py
Comment (sampsyo) [reply mediafile.py:1816]: Interesting! Can you explain a little more about why you want that behavior? (I think this is the way to achieve it, but some testing would be useful, probably.)

File: mediafile.py
Comment (sampsyo) [inline mediafile.py:1647]: It looks like this field doesn't yet have an MP3 mapping?

**Review by sampsyo — COMMENTED** [review]
Thanks for getting this started! I know it's still marked as WIP, but FWIW the tests seem to need some attention (according to CI results).

File: mediafile.py
Comment (JuniorJPDJ) [reply mediafile.py:1647]: You mean I'd need to code separate class for this?

File: mediafile.py
Comment (JuniorJPDJ) [reply mediafile.py:1816]: UPC, UPN and EAN are some specific tags for barcodes which are not supported on most of formats and by most of players.
It would be cool to fill main tag with data from those.

Comment (sampsyo) [reply mediafile.py:1816]: Got it... but I guess what I was asking about was a little more detail on which software *does* support those fields. It seems somewhat harmless to write fields that are unpopular, if that has the effect of increasing compatibility. (That's the usual policy we use in this library.)

File: mediafile.py
Comment (sampsyo) [reply mediafile.py:1647]: Hmm, I wasn't attempting to suggest that. I was just trying to clarify: it looks like the MP3 mapping is disabled for this tag? Is that intentional? Is there an MP3 equivalent to this tag that we *should* be considering, and if so, what do we need to do to support it?

File: mediafile.py
Comment (JuniorJPDJ) [reply mediafile.py]: @sampsyo that's why it's disabled (answer to review below).
I don't know how to handle this.
There's no class like that and I don't know if I need to write mine (would like not to) or if there's any implementation.

File: mediafile.py
Comment (JuniorJPDJ) [reply mediafile.py:1816]: Problem is that you can't easly guess if it's UPC, UPN or EAN, all of them are barcodes but those are different standards for those. UPC and UPN is the same, but EAN is other standard, all of them are BARCODEs tho and should fill main BARCODE tag.

File: mediafile.py
Comment (JuniorJPDJ) [reply mediafile.py:1647]: There's no class like that, it's placeholder ATM.
More at my own review above (I already tagged you).

**Comment by JuniorJPDJ** [comment]
> FWIW the tests seem to need some attention (according to CI results).
I was looking at those (already fixed style check) and it didn't look like it failed on my changes. I need someone to clarify this then and maybe suggest some changes.

File: mediafile.py
Comment (JuniorJPDJ) [reply mediafile.py:1816]: UPC + UPN: https://en.wikipedia.org/wiki/Universal_Product_Code
EAN: https://en.wikipedia.org/wiki/International_Article_Number

File: mediafile.py
Comment (JuniorJPDJ) [reply mediafile.py]: I've more problems:
I mentioned `TXXX:ARTISTS` consisting of array of artists but...
I also have `TXXX:ARTISTS` consisting of one artist but also containing unescaped `\`:
` 'TXXX:ARTISTS': TXXX(encoding=<Encoding.UTF16: 1>, desc='ARTISTS', text=['AC/DC']),`

Comment (JuniorJPDJ) [reply mediafile.py]: This seems to be similar issue to problem I've encountered in #43 
I need some advice about this.

**Comment by JuniorJPDJ** [comment]
`albumartists` is suggested here: https://kodi.wiki/view/Music_tagging#albumartists

**Comment by JuniorJPDJ** [comment]
I think it's ready ;D

**Comment by JuniorJPDJ** [comment]
I rebased changes on top of latest commit.

File: mediafile.py
Comment (JuniorJPDJ) [inline mediafile.py:1820]: I added more tags containing this data and now I'm getting:
```
E   AssertionError: Tags did not match
E     Tag albumartists: ['multival', 'test', 'multival', 'test', 'multival', 'test'] != ['multival', 'test']
```
I'm not sure what's the best way to make it work, maybe someone have some idea?

Comment (JuniorJPDJ) [reply mediafile.py:1820]: Anyway - I'll do new PR with this change later. Now I'm waiting for main code to be merged as my other code is dependent on this PR.

**Comment by sampsyo** [comment]
Thanks for continuing to push this forward! I have to admit that I'm a tiny bit lost in navigating all the changes at this point, so I thought I'd try to sum things up:

* There's new support for storing lists in ID3v2.3 tags.
* The `mb_{album}artistid` fields have become list-valued (depending on the above new feature).
* There are some _other_ brand-new fields (not the above ID fields) that are added.

And of course the tests are updated as well, but that's neither here nor there. Is that an accurate summary?

With apologies for being a picky maintainer, I would love to consider all of these changes individually. It would be great, for example, to carefully consider whether we do actually want to change those ID fields to be lists—unfortunately, I can imagine this change breaking software that depends on their current list behavior. And when we add new fields, we should consider adding them to the documentation. So is there any chance we could break this up into individual steps?

Broadly speaking, adding the new fields should be fairly uncontroversial, although I do want to track down the source of truth (instead of just trusting the HydrogenAudio wiki alone). The IDs-as-lists change could be problematic, however, and needs some careful consideration.

Thanks for considering it!

**Comment by JuniorJPDJ** [comment]
Of course ;)

> The mb_{album}artistid fields have become list-valued (depending on the above new feature).

Will moving this to another MR satisfy you?
Other changes are pretty much depending on themselves so I don't feel like splitting them out is good idea, but if you have other opinion I could do this.

**Comment by JuniorJPDJ** [comment]
I moved multivalue musicbrainz out of here. It's at my local branch atm.

File: mediafile.py
Comment (JuniorJPDJ) [reply mediafile.py]: Moved to #47.

File: test/test_mediafile.py
Comment (sampsyo) [inline test/test_mediafile.py:456]: `not value` is a slightly more Pythonic way to write `len(value) == 0`.

File: mediafile.py
Comment (sampsyo) [inline mediafile.py:841]: Expanding the docstring to cover the new parameters might be nice here.

**Review by sampsyo — COMMENTED** [review]
Awesome; thanks for slimming this PR down!! Is there any chance you could include in the thread here any links to documentation you have about where all these tags came from? You've mentioned the HA wiki and MusicBrainz Picard—does that cover all of these tags, or did you use any other sources? (It's always nice to have documentation for exactly why we used the tag mapping we did.)

File: test/test_mediafile.py
Comment (JuniorJPDJ) [reply test/test_mediafile.py:456]: As I'm already checking a type - you are 100% right, I'll change it :D

File: mediafile.py
Comment (JuniorJPDJ) [reply mediafile.py:841]: It's just an example, this class was already used for other frames in code and now it's more generic so it's possible to use it with most of frames that would need it.
I'm opposed to this as imo it will only make docs less clear.

**Comment by JuniorJPDJ** [comment]
>Is there any chance you could include in the thread here any links to documentation you have about where all these tags came from?

Of course!
I'll do detailed writeup, just gimme an hour or two ;D

**Comment by JuniorJPDJ** [comment]
`artists`: whole thing is defined [in Picard tag mapping](https://picard-docs.musicbrainz.org/downloads/MusicBrainz_Picard_Tag_Map.html)

`copyright`: same as above

`albumartists`: I used [Kodi's scheme](https://kodi.wiki/view/Music_tagging#albumartists) as it's non-standard at the moment and it seems reasonable; ATM adding this tag to Picard is being discussed [here](https://tickets.metabrainz.org/browse/PICARD-700) and [here](https://community.metabrainz.org/t/multiple-album-artists/532302) and I'd like to support all variants, that's why I created #47 (it needs more changes to code); I decided it's better to support main variant first, then fix issues with multivariant and then support all variants

`barcode` changes: I already shown my point of view [in discussion above](https://github.com/beetbox/mediafile/pull/42#discussion_r626086566)

`isrc`: also based on [Picard tag mapping](https://picard-docs.musicbrainz.org/downloads/MusicBrainz_Picard_Tag_Map.html)

`url`: there things are going a bit worse
Most of standards seem not to support URL mapping so I was basing on files tagged by [Tag&Rename](http://www.softpointer.com/tr.htm) some people from my environment were using decoded by `mutagen`
- ID3 is directly supported by `WXXX` frame and it's defined [in spec](https://id3.org/id3v2.4.0-frames)
- MP4 mapping is `©url` and it's based on output of `Tag&Rename` but I wasn't able to find much info about it in google; `ffprobe` decoded it correctly so it seems to be some standard:
```
Input #0, mov,mp4,m4a,3gp,3g2,mj2, from 'kdu4r.mp4':
  Metadata:
    major_brand     : isom
    minor_version   : 512
    compatible_brands: isomiso2avc1mp41
    encoder         : Lavf57.83.100
    compilation     : 0
    URL             : https://www.aliexpress.com/
    media_type      : 0
```
```python
>>> dict(mutagen.File('kdu4r.mp4'))
{'©too': ['Lavf57.83.100'],
 'cpil': False,
 '©url': ['https://www.aliexpress.com/'],
 'furl': ['https://basewin.pl/'],
 'aurl': ['http://chomikuj.pl/'],
 'surl': ['https://drive.google.com/drive/my-drive'],
 'rurl': ['https://ekipatonosi.pl'],
 'burl': ['facebook.com'],
 'stik': [0]}
```
- ASF binding was also got from actual `wma` file tagged by Tag&Rename:
```python
>>> dict(mutagen.File('01. MAROON STRONG.wma'))
{'WM/AudioFileURL': [ASFUnicodeAttribute('https://www.instagram.com/')],
 'Description': [ASFUnicodeAttribute('')],
 'WM/EncodingSettings': [ASFUnicodeAttribute('Lavf58.49.100')],
 'WM/RadioStationURL': [ASFUnicodeAttribute('https://www.netflix.com')],
 'WM/URL': [ASFUnicodeAttribute('https://hostuje.net/')],
 'WM/AuthorURL': [ASFUnicodeAttribute('https://jakdojade.pl')],
 'ID3': [ASFByteArrayAttribute(b'ID3\x03\x00\x00\x00\x00\x00/WXXX\x00\x00\x00\x16\x00\x00\x00\x00https://hostuje.net/COMM\x00\x00\x00\x05\x00\x00\x00eng\x00')],
 'WM/UserWebURL': [ASFUnicodeAttribute('https://hostuje.net/')],
 'WM/MusicBuyUrl': [ASFUnicodeAttribute('https://odrabiamy.pl/')],
 'WM/AudioSourceURL': [ASFUnicodeAttribute('https://www.messenger.com')]}
```
This is however not translated by ffprobe:
```
Input #0, asf, from '01. MAROON STRONG.wma':
  Metadata:
    encoder         : Lavf58.49.100
    WM/URL          : https://hostuje.net/
    WM/UserWebURL   : https://hostuje.net/
    WM/AudioFileURL : https://www.instagram.com/
    WM/AuthorURL    : https://jakdojade.pl
    WM/AudioSourceURL: https://www.messenger.com
    WM/RadioStationURL: https://www.netflix.com
    WM/MusicBuyUrl  : https://odrabiamy.pl/
```
Nor Windows details view show it. Windows however shows another URL tag.
image
This is AuthorURL. But purpose of this tag is other than main URL.

I found some [forum post](https://www.un4seen.com/forum/?topic=2329.25) that also mentioned this tag so I assumed It's widely supported.

- Vorbis/FLAC used by Tag&Rename was `URL` but I added also `WWW` and `Website` as I found actual uses of them in web. This also isn't standarized in any way. EasyTag seems to tag URL input box into `CONTACT` field on flac files.

File: test/test_mediafile.py
Comment (JuniorJPDJ) [reply test/test_mediafile.py]: Fixed!

**Review by sampsyo — COMMENTED** [review]
Thanks for the extra information; this helps a lot!

I admit I'm a bit skeptical of the multiple Vortis-style mappings for `barcode` and the new `url` field. The former is a bit less worrisome because the fields are read-only, but even so, we don't have a lot of evidence that we need to support these mappings. The latter is a bit more worrisome because I know people who use Vorbis comments a lot don't love having duplicate data in their free-form tag data, so multiple mapping should be approached quite carefully.

I know there's no standard approach for these tags, but maybe a good place to start would be to use the most popular mappings that we know are in use in real, frequently used software. Then, we can respond as usage becomes more popular: that is, when someone complains that their other software isn't compatible, we can add a mapping to compensate. What do you think about this more minimalist approach?

File: mediafile.py
Comment (JuniorJPDJ) [reply mediafile.py:841]: Oh god. I had to misread this.
Of course, I'll add docs for parameters. ;)
I don't know why but I had frame types in my mind.

**Comment by JuniorJPDJ** [comment]
> I know people who use Vorbis comments a lot don't love having duplicate data in their free-form tag data

Well.. They won't be happy about `albumartists` from #47 then :X 

I can remove `WWW` and `Website` then, I'm OK with that. Just wanted to maximize compatibility ;)

I'm still struggling with barcode as some standards are not-cool and are defining separate fields for other barcode standards and I honestly would like to have them in one tag, which is understood by most of software.
What do you think, should I remove it or leave it here?

File: mediafile.py
Comment (JuniorJPDJ) [reply mediafile.py:841]:
```python
class MP3DescStorageStyle(MP3StorageStyle):
    """Store data in a TXXX (or similar) ID3 frame. The frame is
    selected based its ``desc`` field.
    ``attr`` allows to specify name of data accessor property in the frame.
    Most of frames uses `text`.
    ``multispec`` specifies if frame data is ``mutagen.id3.MultiSpec`` 
    which means that the data is being packed in the list.
    """
```
What do you think about something like that?

Comment (sampsyo) [reply mediafile.py:841]: Sounds great!! :sparkles:

**Comment by sampsyo** [comment]
Sounds great! For `barcode`: just to clarify, are you saying that there's a trend among software out there to define lots of different fields for different barcode classification standards, such as UPC and EAN?

I can see how that argues in favor of supporting all these miscellaneous tags as read-only mappings and "BARCODE" itself as the only writable version! Googling around reveals threads like [this one](https://hydrogenaud.io/index.php?topic=99290.0) where people are equally frustrated by the proliferation of different tag names for different kinds of barcodes.

Anyway, this sounds perfectly reasonable to me. Happy to do the "read-only for weird names" thing here—until someone out there wants to record _both_ a UPC _and_ an EAN for the same file. 😂

**Comment by JuniorJPDJ** [comment]
> are you saying that there's a trend among software out there to define lots of different fields for different barcode classification standards, such as UPC and EAN?

That's what I try to say. Yes.
Examples are [here](https://wiki.hydrogenaud.io/index.php?title=Tag_Mapping): dBPowerAMP and some wiki for Vorbis tags.

I'll fix my changes in few minutes then and.. It's ready to merge! :D

File: mediafile.py
Comment (JuniorJPDJ) [reply mediafile.py:841]: Done ;D

**Comment by JuniorJPDJ** [comment]
Done! :D

**Comment by JuniorJPDJ** [comment]
> until someone out there wants to record both a UPC and an EAN for the same file.

UPC itself can be converted to EAN by adding 0 at the front, so I think they will not try to kill me for this change even if it overwrite one of them ;)

**Review by sampsyo — APPROVED** [review]
Looks awesome!! Merging now.

**Pull request merged**
